### PR TITLE
turn `append_place_to_string` from recursion into iteration

### PR DESCRIPTION
**Pull request by SparrowLii**
This PR fixes the FIXME in the impl of `append_place_to_string` which turns `append_place_to_string` from recursion into iteration, meanwhile simplifying the code relatively.

### CODE REVIEW

File: compiler/rustc_borrowck/src/diagnostics/mod.rs [hunk 1]
```diff
@@ -175,144 +176,91 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
         place: PlaceRef<'tcx>,
         including_downcast: IncludingDowncast,
     ) -> Option<String> {
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mod.rs:178]: GitHub won't let me comment on the documentation comment for this function, but it is written as if this function returns a `Result`, could you update it to reflect that the function returns `Option`?

File: compiler/rustc_borrowck/src/diagnostics/mod.rs [hunk 2]
```diff
@@ -175,144 +176,91 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
-                        // to avoid confusing the end-user
-                        self.append_place_to_string(
-                            PlaceRef { local, projection: proj_base },
-                            buf,
-                            autoderef,
-                            &including_downcast,
-                        )?;
-                        buf.push_str("[..]");
+                }
+                ProjectionElem::Downcast(..) => {
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mod.rs:226]: nit: could you make this into two branches?

```rust
match elem {
    // ...
    ProjectionElem::Downcast(..) if including_downcast.0 => return None,
    ProjectionElem::Downcast(..) => (),
    // ...
}
```

**Review by davidtwco — CHANGES_REQUESTED** [review]
LGTM, two minor comments

**Comment by SparrowLii** [comment]
have fixed it, thanks!

**Review by davidtwco — APPROVED** [review]

**Pull request merged**